### PR TITLE
[Diagnostic] Fix diagnostic when checking conformance for IUO of gene…

### PR DESCRIPTION
**Pull request by srinikhil-07**
…ric parameter

Prints TypeRepr in diagnostic if possible to throw
accurate diagnostic for IUO conformance checking
Fixes [rdar://problem/64953106]. Fixes SR-13119

The diagnostic emitted when checking conformance for IUO of generic parameter will
print TypeRepr if present to have accurate type information. This prevents 
confusion in the diagnostics emitted.

Fixes [rdar://problem/64953106]. Fixes SR-13119(https://bugs.swift.org/browse/SR-13119).

### CODE REVIEW

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 1]
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
       }
 
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present for IUO
+        TypeLoc typeLoc = TypeLoc();
```
Comment (LucianoPAlmeida) [inline lib/Sema/DerivedConformanceCodable.cpp:362]: I think this can be simplified as `TypeLoc typeLoc;`

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 2]
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
...
+        if(varDecl->isImplicitlyUnwrappedOptional()){
```
Comment (LucianoPAlmeida) [inline lib/Sema/DerivedConformanceCodable.cpp:363]:
```suggestion
        if(varDecl->isImplicitlyUnwrappedOptional()) {
```

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 3]
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
...
+            typeLoc = {
+                varDecl->getTypeReprOrParentPatternTypeRepr(),
+                varDecl->getType(),
+            };
+        } else {
+            typeLoc = {
+                NULL,
+                varDecl->getType(),
+            };
+        }
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceCodable.cpp:373]: Given we're doing this in three places in this function, I think it would be a good idea to extract and reuse:

```c++
static TypeLoc getTypeLocForDiagnostics(VarDecl *VD) {
  auto varTy = VD->getType();
  if (VD->isImplicitlyUnwrappedOptional()) {
    return TypeLoc(VD->getTypeReprOrParentPatternTypeRepr(), varTy);
  }
  return TypeLoc::withoutLoc(varTy);
}
```

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 2 above)
Comment (CodaFi) [reply lib/Sema/DerivedConformanceCodable.cpp:363]: In general, please run `clang-format` on this buffer.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 4]
```diff
@@ -353,8 +358,21 @@ static EnumDecl *synthesizeCodingKeysEnum(DerivedConformance &derived) {
...
+        // for accurate diagnostic, we show TypeRepr if present for IUO
+        TypeLoc typeLoc = TypeLoc();
+        if(varDecl->isImplicitlyUnwrappedOptional()){
+            typeLoc = {
+                varDecl->getTypeReprOrParentPatternTypeRepr(),
+                varDecl->getType(),
+            };
+        } else {
+            typeLoc = {
+                NULL,
```
Comment (CodaFi) [inline lib/Sema/DerivedConformanceCodable.cpp:370]: You don't need to conditionalize this. `getTypeReprOrParentPatternTypeRepr` will always get what the user wrote, if they wrote something at all.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 3 above)
Comment (CodaFi) [reply lib/Sema/DerivedConformanceCodable.cpp:373]: I don't think this needs to be a utility. Its ideal form is four lines

```
TypeLoc typeLoc = {
  varDecl->getTypeReprOrParentPatternTypeRepr(),
 varDecl->getType(),
};
```

**Review by CodaFi — CHANGES_REQUESTED** [review]
A couple of nits, then this is good to merge.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 4 above)
Comment (Nikhil0487) [reply lib/Sema/DerivedConformanceCodable.cpp:370]: @CodaFi 

Yes, but this is breaking nested class/struct types. Please see the test regression I have commented in the https://bugs.swift.org/browse/SR-13119. @varungandhi-apple suggested a change which I inferred to be the above.

Comment (CodaFi) [reply lib/Sema/DerivedConformanceCodable.cpp:370]: Name lookup will see the innermost name in that scope and it should just work out. As long as those fixits compile, this is not a regression and we should update those tests. I very much prefer we preserve the way the user spelled types than fully qualify them all the time.

Comment (Nikhil0487) [reply lib/Sema/DerivedConformanceCodable.cpp:370]: Sure, okay. I get it

File: test/decl/protocol/special/coding/struct_codable_failure_diagnostics.swift [hunk 5]
```diff
@@ -9,8 +9,8 @@ struct S1 : Codable {
   var a: String = ""
   var b: Int = 0
   var c: Nested = Nested()
-  // expected-note@-1 {{cannot automatically synthesize 'Decodable' because 'S1.Nested' does not conform to 'Decodable'}}
-  // expected-note@-2 {{cannot automatically synthesize 'Encodable' because 'S1.Nested' does not conform to 'Encodable'}}
+  // expected-note@-1 {{cannot automatically synthesize 'Decodable' because 'Nested' does not conform to 'Decodable'}}
+  // expected-note@-2 {{cannot automatically synthesize 'Encodable' because 'sNested' does not conform to 'Encodable'}}
```
Comment (LucianoPAlmeida) [inline test/decl/protocol/special/coding/struct_codable_failure_diagnostics.swift:13]: Typo 'sNested' -> 'Nested'?

**Comment by varungandhi-apple** [comment]
Did you try using `Type::getString` with `PrintOptionalAsImplicitlyUnwrapped` set to `true`? Does that avoid the regression for the nested type diagnostics?

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 4 above)
Comment (varungandhi-apple) [reply lib/Sema/DerivedConformanceCodable.cpp]: > I very much prefer we preserve the way the user spelled types than fully qualify them all the time.

I didn't realize there was a disagreement here. Should we write this down somewhere so that newer diagnostics use this style? I thought this was a regression because the original code was printing the full type.

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 6]
```diff
@@ -173,8 +173,13 @@ static bool validateCodingKeysEnum(DerivedConformance &derived,
         break;
 
       case DoesNotConform:
+        // for accurate diagnostic, we show TypeRepr if present
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceCodable.cpp:176]: It might be a good idea to rephrase this to provide better context. It could be something along the lines of:

```suggestion
        // We use a TypeLoc here so diagnostics can show the type
        // as written by the user in source if possible. This is useful
        // when the user has written an IUO type for example, since
        // diagnostics would show the type as 'T?' instead of 'T!' if
        // we use a Type instead.
```

Comment (Nikhil0487) [reply lib/Sema/DerivedConformanceCodable.cpp:176]: Thanks @theblixguy, will update at both the places.

**Review by CodaFi — APPROVED** [review]
Very nice! Thanks ⛵

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 4 above)
Comment (CodaFi) [reply lib/Sema/DerivedConformanceCodable.cpp]: That's a good point. I recall we have a document that tries to explain good diagnostic hygiene. And this is certainly one of those places where you have to be really careful. Because the user-provided type may not always work if you're pulling something from a different DeclContext, and the fully-qualified type may not even work when you have ambiguities caused by e.g. the module-and-type-sharing-names bug.

**Comment by LucianoPAlmeida** [comment]
@Nikhil0487  I think we could update those tests that are failing because although they are misleading because `Int` does conform with codable. They probably will be removed in #32371 anyways ... I can rebase those changes there no problem :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test macOS platform
